cpp/memilio/compartments/flow_model.h

```diff
@@ -0,0 +1,280 @@
+// First a list of tuples is generated for each Tag in Tags, where the tuple is either of type tuple<Tag>, or if
+// Tag == OmittedTag, of type tuple<>. This list is then concatenated, effectively removing OmittedTag.
+template <class OmittedTag, class... Tags>
+decltype(std::tuple_cat(std::declval<typename std::conditional<std::is_same<OmittedTag, Tags>::value, std::tuple<>,
+                                                               std::tuple<Tags>>::type>()...))
+    filter_tuple(std::tuple<Tags...>);
+
+// Function declaration used to replace type T by std::tuple.
+template <template <class...> class T, class... Args>
+std::tuple<Args...> as_tuple(T<Args...>);
+
+// Function declaration used to replace std::tuple by type T.
+template <template <class...> class T, class... Args>
+T<Args...> as_index(std::tuple<Args...>);
+
+// Remove all occurrences of OmittedTag from the types in a std::tuple<types...>.
+template <class OmittedTag, class Tuple>
+using filtered_tuple_t = decltype(filter_tuple<OmittedTag>(std::declval<Tuple>()));
+
+// Remove all occurrences of OmittedTag from the types in an Index = IndexTemplate<types...>.
+template <class OmittedTag, template <class...> class IndexTemplate, class Index>
+using filtered_index_t = decltype(as_index<IndexTemplate>(
+    std::declval<filtered_tuple_t<OmittedTag, decltype(as_tuple(std::declval<Index>()))>>()));
+
+} //namespace details
+
+/**
+ * @brief A FlowModel is a CompartmentalModel defined by the flows between compartments. 
+ *
+ * Inherits from  @see CompartmentalModel, and defines the derivatives depending on the flows. Hence, a model implementing
+ * FlowModel has to define the function get_flows, instead of get_derivatives.
+ *
+ * Flows is expected to be a TypeList containing types Flow<A,B>, where A and B are compartments from the enum Comp.
+ * Some examples can be found in the cpp/models/ directory, within the model.h files.
+ */
+template <class Comp, class Pop, class Params, class Flows>
+struct FlowModel : public CompartmentalModel<Comp, Pop, Params> {
+    using PopIndex = typename Pop::Index;
+    // FlowIndex is the same as PopIndex without the category Comp. It is used as argument type for
+    // get_flat_flow_index, since a flow is used to determine only the compartment (i.e. Index<Comp>) for the
+    // population of the model.
+    // The remaining indices (those contained in FlowIndex) must still be provided to get an entry of population.
+    // This approach only works with exactly one category of type Comp in PopIndex (hence the assertion below).
+    using FlowIndex = details::filtered_index_t<Comp, Index, PopIndex>;
+    // Enforce that Comp is a unique Category of PopIndex, since we use Flows (via their source/target) to provide
+    // Comp indices for the population.
+    static_assert(FlowIndex::size == PopIndex::size - 1, "Compartments must be used exactly once as population index.");
```

Comment:
No, it just guarantees that no category of type Comp is in the FlowIndex. It cannot control how many there where in PopIndex. 
Unless there are checks I am unaware of which guarantee that Comp is used exactly once for the population index, this check is necessary. Otherwise, we'd have to guess which Comp (if there are more than one) to replace with Flow::source/target, which seems like a bad idea to me.

Reply:
that makes sense. I'm pretty sure we don't allow duplicate category entries in populations, since it would break a lot of index lookups. But there probably isn't an explicit check anywhere.